p2p/transport/webrtc/stream.go

```diff
@@ -52,34 +54,42 @@ type sendState uint8
 const (
 	sendStateSending sendState = iota
 	sendStateDataSent
+	sendStateDataReceived
 	sendStateReset
 )
 
 // Package pion detached data channel into a net.Conn
 // and then a network.MuxedStream
 type stream struct {
 	mx sync.Mutex
-	// pbio.Reader is not thread safe,
-	// and while our Read is not promised to be thread safe,
-	// we ourselves internally read from multiple routines...
-	reader pbio.Reader
+
+	// readerOnce ensures that only a single goroutine reads from the reader. Read is not threadsafe
+	// But we may need to read from reader for control messages from a different goroutine.
+	readerOnce chan struct{}
+	reader     pbio.Reader
+
 	// this buffer is limited up to a single message. Reason we need it
 	// is because a reader might read a message midway, and so we need a
 	// wait to buffer that for as long as the remaining part is not (yet) read
 	nextMessage  *pb.Message
 	receiveState receiveState
 
-	// The public Write API is not promised to be thread safe,
-	// but we need to be able to write control messages.
-	writer               pbio.Writer
+	writer               pbio.Writer // concurrent writes prevented by mx
 	sendStateChanged     chan struct{}
 	sendState            sendState
-	controlMsgQueue      []*pb.Message
 	writeDeadline        time.Time
 	writeDeadlineUpdated chan struct{}
 	writeAvailable       chan struct{}
 
-	readLoopOnce sync.Once
+	controlMessageReaderOnce sync.Once
+	// controlMessageReaderEndTime is the end time for reading FIN_ACK from the control
+	// message reader. We cannot rely on SetReadDeadline to do this since that is prone to
+	// race condition where a previous deadline timer fires after the latest call to
+	// SetReadDeadline
+	// See: https://github.com/pion/sctp/pull/290
```

Comment:
Is there anything we can change / simplify here after 290 was merged?

Reply:
This hasn't been merged yet. 


When it's done we can remove the `controlMsgReaderEndTime` and just set the deadline in AsyncClose & Close to be now + 10 seconds. We currently cannot do that because there's no way of knowing whether a deadlineexceeded error occurred because of a past canceled deadline or because of the most recently set deadline.